### ISSUE TITLE
Remove state_record_and_type from erl_files

### ISSUE THREAD

**Issue by elbrujohalcon** [inaka/elvis_core#171]
This is an outdated rule.
The concept is not bad (_When implementing an OTP behavior, define some type to hold its state_), but the actual implementation of the rule requires the type to be called `state/0` and that's too strict. For instance, we have implementations of `gen_statem` where it makes much more sense to call that type `data/0` or `state_data/0`.

We should either allow the record name to be matched against a more flexible regular expression or just remove the rule from `erl_files` altogether.

**Comment by paulo-ferraz-oliveira** [comment]
We could have a reasonable default per OTP behaviour, though this will be an ever-growing list, I imagine.

e.g.

```erlang
#{ gen_server => state
 , gen_statem => [data, state_data]
 }
```

Removing it might also be reasonable (to not break interface, though, we should issue a warning without changing the exit code).

**Comment by paulo-ferraz-oliveira** [comment]
> call that type data/0 or state_data/0.

👍, we usually call it `sdata`, even.

**Comment by paulo-ferraz-oliveira** [comment]
Marking this is `bug` since it's a "very desirable" update.

**Comment by tsloughter** [comment]
Just upgraded `rebar3_lint` on Elli and this warning popped up which led me to find it doesn't support if the type is `opaque` https://github.com/elli-lib/elli/blob/main/src/elli.erl#L52

It should probably at least not warn if the type is there but `-opaque` instead of `-type`?

**Comment by elbrujohalcon** [comment]
@tsloughter you are right. It shouldn't warn on `-opaque` but, why are you exporting the type of your internal state? 🤔

**Comment by paulo-ferraz-oliveira** [comment]
@elbrujohalcon, do you think we should fix the rule or rather remove it, as proposed by the issue itself?

**Comment by paulo-ferraz-oliveira** [comment]
But keep the rule itself?

You mean, remove it from the default, right?

**Comment by elbrujohalcon** [comment]
Right. Remove it from the erl_files ruleset only

**Comment by paulo-ferraz-oliveira** [comment]
On it!

**Comment by paulo-ferraz-oliveira** [comment]
Sorry for side-tracking this: I've found that `no_behavior_info` isn't applicable (by default) to BEAM files. I'm not sure how it would behave (would need to test), but the Wiki says it works. Should we also include it in the default?

**Comment by elbrujohalcon** [comment]
Yeah, I think we should.

**Comment by elbrujohalcon** [comment]
We should not include it… As a matter of fact, it doesn't _really_ work for BEAM files. Since that function (`behaviour_info`) is auto-included by the compiler… it will warn for every single behaviour definition out there and it won't be possible to silence it.

**Comment by elbrujohalcon** [comment]
The real solution is to state that it doesn't work for BEAM files in the Wiki.
